### DISCUSSION TITLE
Fitting covariates in moveHMM

### DISCUSSION THREAD

**Discussion by eelger** [TheoMichelot/moveHMM#41] [Q&A] [answered]
Hi, I'm working with an animal movement data set and I have been fitting HMM with moveHMM. I am coming across a problem when I try to fit time as a covariate. I have tried several permutations of fitting time in the model (including converting time to numerical 1-24 and testing it on chunks of the data where there is no missinggness). However, when I have fitted the model with time added (seems to work successfully) and I try to run the plot() function I keep getting the error: "Error in if (maxdens > ymax & maxdens < 1.5 * ymax) { :  missing value where TRUE/FALSE needed".  My use of moveHMM thus far follows the moveHMM wild haggis example and it has been working very smoothly with all commands up until this point. Many thanks :)

**Comment by TheoMichelot** [comment, marked as answer]
This error message can happen for a number of reasons, but usually it means that the model didn't fit correctly. When there are numerical problems during model fitting, some parameters might be estimated to take very large or unrealistic values, which can cause that error. 

There is no one-size-fits-all solution, but I would recommend (1) trying a simpler model to start with (without covariates), and (2) possibly trying different initial parameter values (stepPar0 and anglePar0). For (2), see the vignette: [Choosing starting values in moveHMM](https://cran.r-project.org/web/packages/moveHMM/vignettes/moveHMM-starting-values.pdf). A covariate-free model is less likely to run into numerical problems, and parameter estimates from that model can be used as starting values for a more complex model with effect of time.

It sounds like your time variable is the time of day, i.e., a number between 0 and 24. If so, you should also model the effect with a cyclical function, to ensure that the effect at 0 matches the effect at 24. The most common way to do this is to use trigonometric functions, and the formula would look something like:
``` R
formula <- ~ cos(2 * pi * tod / 24) + sin(2 * pi * tod / 24)
```
where `tod` is the time of day variable. If `tod` is measured in hours, the above formula ensures a 24-hour cycle.
